File: src/simulation/navigation/Faction.js
// v0.8.4 strategic walls + GATE — Agent C
//
// Faction-aware passability. Walls and gates have always been on the map but
// pre-v0.8.4 they blocked everyone symmetrically — workers couldn't enter
// their own walled compound, raiders couldn't either. The gate tile fixes
// the colony side: gates are passable for the colony faction and blocked
// for hostile/neutral factions. Walls remain impassable for everyone.
//
// `getEntityFaction(entity)` derives a stable faction tag from the
// canonical entity fields:
//   - WORKER (any role, including GUARD)               → "colony"
//   - VISITOR (TRADER subtype)                          → "colony"
//   - VISITOR (SABOTEUR subtype)                        → "hostile"
//   - ANIMAL  (HERBIVORE)                               → "neutral"
//   - ANIMAL  (PREDATOR, incl. raider_beast)            → "hostile"
//   - any unknown shape                                 → "colony" (safe default)
//
// `isTilePassableForFaction(tileType, faction)` is the gate the AStar
// neighbour-loop calls. It enforces:
//   - WALL: never passable for any faction.
//   - GATE: passable only for "colony" faction.
//   - WATER: always blocked (delegated to TILE_INFO.passable, which is
//     `false`). The caller still consults TILE_INFO.passable separately
//     for non-faction-aware blocks; this helper layers faction checks on
//     top.
//
// Wiring lives in:
//   - AStar.js: passes options.faction (default "colony") through the
//     neighbour loop.
//   - Navigation.js: derives faction via getEntityFaction(entity) and
//     threads it into AStar + PathCache + worker pool.
//   - PathCache.js: caches paths keyed by faction so colony/hostile
//     searches never collide.
//   - PathWorkerPool.js / pathWorker.js: faction propagated through
//     payloads.
//
// See docs/superpowers/plans/2026-04-28-building-construction-walls-recruit.md
// § 4 for the full design contract.
import { TILE } from "../../config/constants.js";

export const FACTION = Object.freeze({
  COLONY: "colony",
  HOSTILE: "hostile",
  NEUTRAL: "neutral",
});

/**
 * @param {object|null|undefined} entity
 * @returns {"colony" | "hostile" | "neutral"}
 */
export function getEntityFaction(entity) {
  if (!entity) return FACTION.COLONY;
  const entityType = String(entity.type ?? "");
  if (entityType === "WORKER") return FACTION.COLONY;
  if (entityType === "VISITOR") {
    const kind = String(entity.kind ?? "");
    return kind === "SABOTEUR" ? FACTION.HOSTILE : FACTION.COLONY;
  }
  if (entityType === "ANIMAL") {
    const kind = String(entity.kind ?? "");
    return kind === "PREDATOR" ? FACTION.HOSTILE : FACTION.NEUTRAL;
  }
  return FACTION.COLONY;
}

/**
 * Faction-aware passability check. WALL is always blocked. GATE is only
 * passable for the colony faction. All other tiles defer to the caller's
 * `TILE_INFO.passable` check (for things like water).
 *
 * Note: callers should still consult `TILE_INFO.passable` first for
 * non-faction blocks (water, etc.). This helper only encodes the
 * faction-specific delta.
 *
 * @param {number} tileType
 * @param {string} faction — typically the result of `getEntityFaction`
 * @returns {boolean}
 */
export function isTilePassableForFaction(tileType, faction) {
  if (tileType === TILE.WALL) return false;
  if (tileType === TILE.GATE) return faction === FACTION.COLONY;
  return true;
}

File: project_utopia/simulation/navigation/faction.py
"""Faction-aware tile passability (port of ``src/simulation/navigation/Faction.js``).

Walls block everyone; gates are passable for the ``COLONY`` faction only.
All other faction filters delegate to ``TILE_INFO.passable`` (handled in
the A* neighbour loop).

Faction tags:
    COLONY   — workers, traders
    HOSTILE  — saboteurs, predators
    NEUTRAL  — herbivores
"""

from __future__ import annotations

from types import MappingProxyType
from typing import Any

from project_utopia.world.grid import TILE

__all__ = [
    "FACTION",
    "get_entity_faction",
    "is_tile_passable_for_faction",
]

# Frozen faction enum. ``MappingProxyType`` gives a read-only dict view,
# mirroring JS ``Object.freeze``.
FACTION: MappingProxyType[str, str] = MappingProxyType(
    {
        "COLONY": "colony",
        "HOSTILE": "hostile",
        "NEUTRAL": "neutral",
    }
)


def get_entity_faction(entity: Any) -> str:
    """Derive a faction tag from an entity-like object.

    Handles both dataclass-style attribute access (``entity.type``) and
    dict-style access (``entity["type"]``). Returns ``COLONY`` as the safe
    default for unknown shapes (matches the JS behaviour).

    Routing rules:
        WORKER (any role)                   → colony
        VISITOR/TRADER                      → colony
        VISITOR/SABOTEUR                    → hostile
        ANIMAL/HERBIVORE                    → neutral
        ANIMAL/PREDATOR (incl. raider)      → hostile
        anything else                       → colony (safe default)
    """
    if entity is None:
        return FACTION["COLONY"]

    entity_type = _get(entity, "type", "")
    if entity_type == "WORKER":
        return FACTION["COLONY"]
    if entity_type == "VISITOR":
        kind = _get(entity, "kind", "")
        return FACTION["HOSTILE"] if kind == "SABOTEUR" else FACTION["COLONY"]
    if entity_type == "ANIMAL":
        kind = _get(entity, "kind", "")
        return FACTION["HOSTILE"] if kind == "PREDATOR" else FACTION["NEUTRAL"]
    return FACTION["COLONY"]


def is_tile_passable_for_faction(tile_type: int, faction: str) -> bool:
    """Return ``True`` iff ``faction`` may cross ``tile_type``.

    Layered semantics: callers should consult ``TILE_INFO.passable`` first
    (which already blocks WATER, WALL for everyone). This helper only
    encodes the faction-specific delta — i.e. gates close to non-colony
    factions and walls are double-checked as a safety net.
    """
    if tile_type == TILE["WALL"]:
        return False
    if tile_type == TILE["GATE"]:
        return faction == FACTION["COLONY"]
    return True


def _get(entity: Any, key: str, default: Any = None) -> Any:
    """Read ``entity[key]`` whether ``entity`` is a dict or an object."""
    if isinstance(entity, dict):
        return entity.get(key, default)
    return getattr(entity, key, default)
